### ISSUE TITLE
[Suggestion] Configurable Speed-upgrade ratios

### ISSUE THREAD

**Issue by accessdenied0** [turbodiesel4598/NuclearCraft#248]
Hey NuclearCraft team.

I really like this mod, played with it since 1.7.10, to this time there were Energy-Upgrades a thing.

Well, I like what you do with your mod but I would love to have a bit of control over the Speed-upgrade ratios. Especially they get insanely expensive really fast, make it for most utilisation in compare to with other mods really hard/expensive.

On the other hand some of the machines need these Ratios....

tl;dr: I would love to have an option, which allows me to turn down/up the energy/speed multiplier.

If this feature already exists (which I didn't found), I'm deeply sorry.

I really look forward in the future of this mod! Thank you for maintaining and all your effort.

access_denied

**Comment by turbodiesel4598** [comment]
Some say there's no I in team, but in my case, I suppose there is :)

When you say turn down/up the multiplier, I guess you mean the base percentage change, right?

**Comment by accessdenied0** [comment]
Oh ok, that makes your work even more impressive.

Yes I meant the percentage changes like  X "Power-consumed" per Y "Speed-Multiplier".

It would be great to be able to adjust X and Y.

I hope that is in your interest for this mod.

**Comment by Vyraal1** [comment]
Also think this is a great idea (although I think your default values are also just fine). Could have the configs line be a double or whatever and so if I wanted each upgrade to increase energy usage by 4x while only increasing speed by 2x, it'll be:
```
Power Multiplier: 200
Speed Multiplier: 400
```

**Comment by turbodiesel4598** [comment]
The reason I say 'base' change is because the speed multiplier is linear in the number of upgrades while the power multiplier is quadratic. So do you want an additional config option to change the power laws of the upgrades, as well as the config option to change the effect of one upgrade?

**Comment by accessdenied0** [comment]
Mh.. ok... Can I get the math to for it?

You mean like Speed^Energy? How do you calculate the Energy/Speed?

**Comment by turbodiesel4598** [comment]
The speed upgrade directly affects the power use and the time per operation, which both in turn implicitly affect the energy per operation (energy = power/speed).

By default, if **n** is the number of speed upgrades installed, then the power multiplier is **n×(n+1)/2** and the speed multiplier is **n**, resulting in an energy multiplier of **(n+1)/2**.

Both of these expressions are particular forms of the general expression **[n+p]C[p]**, which can also be written as **(n+p)!/(n!×p!)**, where **p** is the power law. For the power multiplier, **p=2**, while for the speed, **p=1**. I could add a config option that allows these values of **p** to be altered.

**Comment by accessdenied0** [comment]
ahhh... so it is technically like **[f(x,p) = (n+p)!/(n!+p!)](http://puu.sh/zKfaE/88934e4694.png)** while **p** either **1 or 2** is depending on either it is **power (p = 2)** or **speed (p = 1)**. So Speed graph is linear and the Power is "quadratic"... Now I get it.

Yeah I think being able to adjust both variables are great. To make it like adjust the grade of the graph means the slope of the graph, is to much work, which I can't ask for.

**Comment by turbodiesel4598** [comment]
Yep, exactly. Okay, I'll add two config options then, for the power laws of the power use and speed :)

**Comment by accessdenied0** [comment]
Thank you very much for your work and the explaining of your work, too.

**Comment by turbodiesel4598** [comment]
No prob!

**Comment by accessdenied0** [comment]
One last question, because now that I think of... 
I have to apologise, because I feel slightly broadminded...
How exactly do you add the multiplier? 
I came of the idea of an after modifier m :  [f(x)=[!(x+p)!/(x!*p!)]/m](http://puu.sh/zKfQC/1c4a864014.png) **m (element of) [0.05,∞[**

Wouldn't it technically adjust the slope and with it the ratio of the values?

Like I said I just ask because I feel slightly dumb right now ^^'

**Comment by accessdenied0** [comment]
I think I see the problem... if m is in the calculation of p =1 very directional, it is in p = 2 inverted exponential....

**Comment by turbodiesel4598** [comment]
A modifier like that would work I think - it would just be an overall multiplier, so it would tell you how big an effect a single upgrade would have.

**Comment by accessdenied0** [comment]
How would you make it? I might have a better view to it, because so far I was the impression both calculations where made separately:
Speed = (x+p)!/(x!*p!)
Power = (x+p)!/(x!*p!)
x is unchangeable because it determines the count of the Upgrades, means the formula is only adjustable by changing p and with it he grade of the function or by adding another variable to the term, which could be done by two multiplier:
Speed = (config.speed)^p * (x+p)!/(x!*p!)
Power = (config.power)^p * (x+p)!/(x!*p!)
Or do you wanna avoid this?

**Comment by turbodiesel4598** [comment]
Yeh, there would be a separate multiplier for both the speed and the power, so four config options in total.

**Comment by accessdenied0** [comment]
Thank you so much! You do a great job and put a lot of effort in your mod and videos.
